[NuGet] Support conditional PackageReferences when multi-targeting

mrward opened the pull request:
Use the GenerateRestoreGraphFile MSBuild target to deterine the
package spec information instead of determining this from the project
model. This allows multi-targeting projects and conditional
PackageReferences to be supported.

Fixes VSTS #583482 - PackageReference does not respect condition
Fixes VSTS #572320 - Support conditional NuGet PackageReferences in
multi-targeting projects
Fixes VSTS #572329 - Show per-framework dependencies when multi-targeting
Fixes VSTS #563083 - Using Paket with new .fsproj format
Fixes VSTS #738993 - Type system not updated after saving edited project file in editor
Fixes VSTS #714265 - NuGet restores the wrong version of Microsoft.AspNetCore.App
Fixes VSTS #711758 - NuGet restore ignore build targets
Fixes VSTS #740214 - NuGet packages are not getting restored for ASP.NET Core projects with error "System.NullReferenceException"

mrward committed:
[NuGet] Support conditional PackageReferences when multi-targeting

Use the GenerateRestoreGraphFile MSBuild target to deterine the
package spec information instead of determining this from the project
model. This allows multi-targeting projects and conditional
PackageReferences to be supported.

Fixes VSTS #583482 - PackageReference does not respect condition
Fixes VSTS #572320 - Support conditional NuGet PackageReferences in
multi-targeting projects
Fixes VSTS #572329 - Show per-framework dependencies when multi-targeting
Fixes VSTS #563083 - Using Paket with new .fsproj format

mrward committed:
[NuGet] Allow verbose output when getting restore info from msbuild

Added an environment variable which will show the msbuild output
when the GenerateRestoreGraphFile target is run in the Package Console
window.

Also log a message when the restore info is being obtained from the
GenerateRestoreGraphFile target to the Package Console. If this
takes some time nothing was being displayed in the Package Console
window.

mrward committed:
Revert "[DotNetCore] Handle implicit package reference versions"

Workaround to handle implicit package references is no longer
required. The package versions are not found using the
GenerateRestoreGraphFile MSBuild target.

This reverts commit 8e42df53e59fabb261933cd9280572404d28b83b.

mrward committed:
[NuGet] Non sdk style projects use MSBuild target to get package info

Non SDK style PackageReference projects now use the
GenerateRestoreGraphFile MSBuild target to get package information.

mrward committed:
[NuGet] Remove unused code

The PackageSpecCreator is no longer used. Replaced by the
MSBuildPackageSpecCreator which uses the GenerateRestoreGraphFile
MSBuild target to get PackageReference information.

mrward committed:
[NuGet] Improve solution restore performance

For solutions that contain many projects generating the restore
information for each project can be slow compared with getting
it in one step. The solution restore now uses similar approach to
what nuget.exe uses. MSBuild is run with a generated project file
that includes the PackageReference projects that are to be restored.
This allows the package spec information to be generated in one step
which is faster than using the msbuild host and getting the package
information for each project, when each project depends on several
other projects, such as with OrchardCore.

Timings: OrchardCore - git clean before each test

Package restore info - one project: 2 mins.
Package restore info - solution: 30 seconds

Therzok commented on main/src/addins/MonoDevelop.PackageManagement/MonoDevelop.PackageManagement/MSBuildPackageSpecCreator.cs:
```diff
@@ -0,0 +1,92 @@
+
+		public static DependencyGraphSpec GetDependencyGraph (string resultsPath)
+		{
+			if (File.Exists (resultsPath) && new FileInfo (resultsPath).Length != 0) {
```
We should use the same file info:
```suggestion
			var fileInfo = new FileInfo (resultsPath);
			if (fileInfo.Exists && fileInfo.Length != 0) {
```

Therzok commented on main/src/addins/MonoDevelop.PackageManagement/NuGet.Commands/MSBuildUtility.cs:
```diff
@@ -0,0 +1,197 @@
+						arguments,
+						solution.BaseDirectory,
+						monitor.Log,
+						monitor.Log,
```
Shouldn't this be ErrorLog?

Therzok commented on main/src/addins/MonoDevelop.PackageManagement/NuGet.Commands/MSBuildUtility.cs:
```diff
@@ -0,0 +1,197 @@
+						arguments,
+						solution.BaseDirectory,
+						monitor.Log,
+						monitor.Log,
```
```suggestion
						monitor.ErrorLog,
```

mrward replied:
I think I did this after looking at [existing code](https://github.com/mono/monodevelop/blob/0a95558a7da163a86a6887b0dfbe78292b4f4283/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/CustomCommand.cs#L301) which also does not use the ErrorLog.

Therzok replied:
That code should also be changed.

Therzok commented on main/src/addins/MonoDevelop.PackageManagement/NuGet.ProjectManagement/TempFile.cs:
```diff
@@ -0,0 +1,71 @@
+			do {
+				filePath = Path.Combine (tempDirectory, Path.GetRandomFileName () + extension);
+
+				if (!File.Exists (filePath)) {
```
We retry 3 times if the file exists?

Therzok committed:
[NuGet] Use progress monitor's ErrorLog on restore

The same progress monitor Log property was being passed
to ProcessService.StartProcess.

Co-Authored-By: mrward <ward.matt@gmail.com>

mrward committed:
[NuGet] Cache package spec information

If a project is restored using the MSBuild host and it has project
dependencies then the package spec information is created for
these referenced projects. This package spec information is added
to the DependencyGraphContext. Previously only the project being
restored was added which could result in the GenerateRestoreGraphFile
target being run again for information that had already been
obtained.

There is still some overlap here if two projects share dependencies
but do not reference each other. This may still result in a restore
running for the same project dependencies. For projects that do not
have a large amount of dependencies this is OK and is still faster
than running MSBuild externally.

mrward committed:
[NuGet] Refactor MSBuildPackageSpecCreator

Use the FileInfo to check for a file's existence and length.

mrward committed:
[NuGet] Reverse topological sort projects for restore

This improves the performance when restoring using the MSBuild host
since a project with project references will include package info
from the GenerateRestoreGraphFile MSBuild target for all referenced
projects. These are cached and avoids some calls to this MSBuild
target.

mrward committed:
[NuGet] Improve performance when restoring smaller solutions

Use the MSBuild host to get the dependency graph when restoring
solutions that have 10 or fewer PackageReference projects. Getting the
dependency graph information for more than 10 projects individually
using the MSBuild host starts to take about the same time as running
msbuild out of process for the solution. Another assumption is that
larger solutions may have more complex dependency graphs, such as
OrchardCore, where using the MSBuild host can be much slower since
it may recompute the same information based on root projects in
the dependency tree.

mrward committed:
[Core] Out of date project builder used on reload

Editing a .csproj file to change a reference and after the project
was reloaded the type system would still be using the old reference
information. The problem was that the type system service reloads
the project when the reloaded project is added to the solution.
This can happen before the old project is disposed, which shuts down
the project builder, resulting in the old out of date project builder
being used. To try to prevent this the project is unloaded from the
remote build engine before the SolutionItemAdded event is fired.
A new project builder should then be created when the type system
gets the reference information for the project.

Fixes VSTS #738993 - Type system not updated after saving edited
project file in editor

mrward commented:
Reworked this pull request with some performance improvements. Also fixes a problem with the type system using out of date reference information on project reload.

slluis reviewed (changes requested):
Two general comments:
1) There are several references to `IdeApp.Workspace.ActiveConfiguration`. The changes you are doing are on the project system, and the project system should not depend on the IDE status. If the configuration is required for getting package information, it should be provided as an argument.
2) There is logic in MSBuildUtility for running msbuild, which involves locating the right mono, msbuild, building the command line, etc. I'd like this logic to be encapsulated in MonoDevelop.Projects.MSBuild, since MSBuildProjectService uses a similar logic.

mrward committed:
[Core] Move MSBuild process runner logic out of NuGet addin into Core

Added an MSBuildProcessService which can be used to directly run
msbuild.

mrward commented:
Moved part of the code out of MSBuildUtility into a new MSBuildProcessService class in the MonoDevelop.Projects.MSBuild project.

Not sure what to do about the IdeApp.Workspace?.ActiveConfiguration. This is used in the NuGet addin, not directly in the project system, and would be based on the current setting in the IDE. I could push this up to higher in the stack if that would be better.

slluis commented:
If the code above that uses IdeApp.Workspace ends being called directly or indirectly by the project system, then the configuration must be a parameter that has to be provided by the project system. The case I mean is for example, a project model extension invoking that code to find referenced assemblies or something like that. The project system has no concept of active configuration. An invocation to a method is supposed to always have the same effect, no matter what is the configuration selected in the IDE.

If the above code is only invoked by the NuGet service, then maybe the reference may stay. However, you have to be aware (and users of the api must be aware) that the result of the functions may change depending on what the user selects in the IDE, and the service should probably react to active configuration changes.

mrward commented:
When you restore or install a new NuGet packages the current configuration would be used at that time - it is not captured forever. So in the unusual case where the user has package references that are conditional on the platform/configuration they would need to restore the packages to fix the references. I can test this in VS on Windows - I suspect they do not automatically restore on switching configuration.

mrward commented:
VS 2017 runs a restore for a .NET Core project when the configuration is switched between Debug/Release even if there are no conditions for the PackageReferences. VS Mac can do the same.

mrward committed:
[Core] Fix project builder null ref after project reload

If the project is reloaded or disposed whilst an MSBuild target
is being run by the project builder this can result in the next
MSBuild target using the same project builder, but failing to
increase its reference count since it will be shutdown. Then
a null reference exception can be thrown if the build engine
is used since it will be set to null on disposing the builder.

Fixes VSTS #740214 - NuGet packages are not getting restored for
ASP.NET Core projects with error "System.NullReferenceException"

mrward committed:
[NuGet] Remove duplicate code

The MonoDevelopSolutionManager project cache was being cleared
twice. The PackageManagementWorkspace clears the cache for the
same solution item added/removed events so there is no need for the
solution manager to also do this.

mrward committed:
[NuGet] Pass configuration to MSBuild restorers

Moved the configuration logic up to the PackageManagementWorkspace
and MonoDevelopSolutionManager classes. The active configuration
is used when restoring packages. This can be used to have different
PackageReferences for a particular configuration. Previously this
configuration was captured in the msbuild restore classes that
ran the restore targets.

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/RemoteBuildEngine.cs:
```diff
@@ -153,8 +153,7 @@ async Task<int> LoadProject (string projectFile)
 		{
-			lock (remoteProjectBuilders)
-				remoteProjectBuilders.Remove (remoteBuilder.File);
+			OnShuttingDownProjectBuilder (remoteBuilder);
```
I don't quite understand this commit. If a builder is being shut down it doesn't qualify for being reused for running another target. I don't see in which case the builder could still be in use when UnloadProject is called.

mrward replied:
Correct - if UnloadProject is called then the project builder will not be available. The problem is that if there is reference held for the project builder, so it is still in use, then UnloadProject will not be called until the project builder has finished running the target. So a project builder can have its Shutdown method called but UnloadProject will not be called since it is still running. This can result in the project builder being re-used in [GetRemoteProjectBuilder](https://github.com/mono/monodevelop/blob/b7f59d3c01662a66cae069e6dca66ad9e34cb7d4/main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/RemoteBuildEngine.cs#L117) since it is still in the list even though it has been marked as shutting down.

 I added a unit test which reproduced this problem before fixing the bug.

slluis replied:
I think the real problem is that `GetRemoteProjectBuilder()` doesn't take into account that `RemoteProjectBuilder.AddReference()` will fail if the builder is being shut down. The fix would be for `GetRemoteProjectBuilder()` to check the result of `AddReference` before returning a builder, and if it fails, unlist the builder and create a new one.

mrward replied:
Reworked the code so the AddReference method is checked in GetRemoteProjectBuilder.

mrward committed:
[Core] Handle builder shutting down in engine's GetRemoteProjectBuilder

Reworked the logic to handle a project builder shutting down inside
the RemoteBuildEngine's GetRemoteProjectBuilder instead of having
the project builder remove itself from the list of builders when
ReleaseReference is called. The build engine checks the return value
of AddReference and if this false, which means the builder is shutting
down, it is removed from the list of builders and a new one is
created.

slluis commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/RemoteBuildEngine.cs:
```diff
@@ -115,9 +115,16 @@ public Task<RemoteProjectBuilder> GetRemoteProjectBuilder (string projectFile, b
+						else // Project builder is shutting down
+							remoteProjectBuilders.Remove (projectFile);
+					} else if (addReference) {
 						builder.ContinueWith (t => t.Result.AddReference (), TaskContinuationOptions.NotOnFaulted);
```
AddReference should also be checked here, since by the time this continuation is executed, the builder may already have been shut down.

mrward committed:
[Core] Check builder referenced in continuation

Check the project builder could be referenced in the Task
continuation and if not remove it from the list of builders.
